projects/l2_pooling/noise_tolerance.py

```diff
@@ -0,0 +1,809 @@
+    # once more, testing it.
+    numTouchesPerTraversal = len(featureLocations) / float(numColumns)
+    numInitialTouches = int(math.ceil(numInitialTraversals * numTouchesPerTraversal))
+    numTestTouches = int(math.ceil(1 * numTouchesPerTraversal))
+    for touch in xrange(numInitialTouches + numTestTouches):
+      sensorPositions = next(sensorPositionsIterator)
+      for _ in xrange(3):
+        allLateralInputs = [pooler.getActiveCells() for pooler in poolers]
+        for columnNumber, pooler in enumerate(poolers):
+          noiseLevel = max(0.0, min(1.0, noiseFn()))
+
+          position = sensorPositions[columnNumber]
+          featureLocationName = featureLocations[position]
+          feedforwardInput = featureLocationSDRs[columnNumber][featureLocationName]
+          feedforwardInput = noisy(feedforwardInput, noiseLevel, L4_CELL_COUNT)
+
+          lateralInputs = [lateralInput
+                           for i, lateralInput in enumerate(allLateralInputs)
+                           if i != columnNumber]
+
+          pooler.compute(feedforwardInput, lateralInputs, learn=False)
+
+      if touch >= numInitialTouches:
+        for columnNumber, pooler in enumerate(poolers):
+          activeCells = set(pooler.getActiveCells())
+          inferredCells = objectL2Representations[objectName][columnNumber]
+
+          results.append((len(activeCells & inferredCells),
+                          len(activeCells - inferredCells)))
+
+  return results
+
+
+def doConstantNoiseExperiment(numColumns, l2Overrides, objectDescriptions,
+                              noiseLevel, numInitialTraversals):
+
+  noiseFn = lambda: noiseLevel
+
+  return doExperiment(numColumns, l2Overrides, objectDescriptions, noiseFn,
+                      numInitialTraversals)
+
+
+def doVaryingNoiseExperiment(numColumns, l2Overrides, objectDescriptions,
+                             noiseMu, noiseSigma, numInitialTraversals):
+
+  noiseFn = lambda: random.gauss(noiseMu, noiseSigma)
+
+  return doExperiment(numColumns, l2Overrides, objectDescriptions, noiseFn,
+                      numInitialTraversals)
+
+
+def constantNoise_varyColumns():
+  #
+  # Run the experiment
+  #
+  noiseLevels = [x * 0.01 for x in xrange(0, 101, 5)]
+  l2Overrides = {"sampleSizeDistal": 20}
+  columnCounts = [1, 2, 3, 4]
+
+  results = defaultdict(list)
+
+  for trial in xrange(1):
+    print "trial", trial
+    objectDescriptions = createRandomObjectDescriptions(10, 10)
+
+    for numColumns in columnCounts:
+      print "numColumns", numColumns
+      for noiseLevel in noiseLevels:
+        r = doConstantNoiseExperiment(numColumns, l2Overrides,
+                                      objectDescriptions, noiseLevel,
+                                      numInitialTraversals=6)
+        results[(numColumns, noiseLevel)].extend(r)
+
+  #
+  # Plot it
+  #
+  numCorrectActiveThreshold = 30
+  numIncorrectActiveThreshold = 10
```

Comment:
I prefer to keep them local and flexible. There's no correct value for them, different experiments might choose to use different thresholds. I like for the code to imply that.